lib/rubygems/printable_uri.rb

```diff
@@ -0,0 +1,81 @@
+# frozen_string_literal: true
+
+require 'uri'
+require_relative 'uri_parser'
+
+class Gem::PrintableUri
+  def self.parse_uri(uri)
+    new(uri).parse_uri
+  end
+
+  def initialize(original_uri)
+    @credential_redacted = false
+    @original_uri = original_uri
+  end
+
+  def parse_uri
+    @original_uri = Gem::UriParser.parse_uri(@original_uri)
+    @uri = @original_uri.clone
+    redact_credential
+
+    self
+  end
+
+  def parsed_uri?
+    @uri.is_a? URI::Generic
+  end
+
+  def credential_redacted?
+    @credential_redacted
+  end
+
+  def original_password
+    return unless parsed_uri?
+
+    @original_uri.password
+  end
+
+  def to_s
+    @uri.to_s
+  end
+
+  private
+
+  def redact_credential
+    return unless redactable_credential?
+
+    if token?
+      @uri.user = 'REDACTED'
+    elsif oauth_basic?
+      @uri.user = 'REDACTED'
+    elsif password?
+      @uri.password = 'REDACTED' if password?
+    end
+
+    @credential_redacted = true
+  end
+
+  def redactable_credential?
+    return false unless parsed_uri?
```

Comment:
All these checks are maybe a bit too defensive? Maybe we can avoid calling `redact_credential` at all unless `parsed_uri?` and remove all these checks assuming we're always dealing with proper URIs. By the way, this makes me think that maybe `parsed_uri?` should be named `valid_uri?`. As far as I understand, if `GemParser.parse_uri` returns the original string, it means it was not given a valid uri, so we cannot further look for passwords or usernames, correct?